src/librustc_passes/intrinsicck.rs

```diff
@@ -119,6 +122,262 @@ impl ExprVisitor<'tcx> {
+        // possibly fail is for SIMD types which don't #[derive(Copy)].
+        if !ty.is_copy_modulo_regions(self.tcx, self.param_env, DUMMY_SP) {
+            let msg = "arguments for inline assembly must be copyable";
+            let mut err = self.tcx.sess.struct_span_err(expr.span, msg);
+            err.note(&format!("`{}` does not implement the Copy trait", ty));
+            err.emit();
+        }
+
+        // Ideally we wouldn't need to do this, but LLVM's register allocator
+        // really doesn't like it when tied operands have different types.
+        //
+        // This is purely an LLVM limitation, but we have to live with it since
+        // there is no way to hide this with implicit conversions.
+        //
+        // For the purposes of this check we only look at the `InlineAsmType`,
+        // which means that pointers and integers are treated as identical (modulo
+        // size).
+        if let Some((in_expr, Some(in_asm_ty))) = tied_input {
+            if in_asm_ty != asm_ty {
+                let msg = &format!("incompatible types for asm inout argument");
+                let mut err = self.tcx.sess.struct_span_err(vec![in_expr.span, expr.span], msg);
+                err.span_label(
+                    in_expr.span,
+                    &format!("type `{}`", self.tables.expr_ty_adjusted(in_expr)),
+                );
+                err.span_label(expr.span, &format!("type `{}`", ty));
+                err.note("asm inout arguments must have the same type");
+                err.note("unless they are both pointers or integers of the same size");
+                err.emit();
+            }
+
+            // All of the later checks have already been done on the input, so
+            // let's not emit errors and warnings twice.
+            return Some(asm_ty);
+        }
+
+        // Check the type against the list of types supported by the selected
+        // register class.
+        let asm_arch = self.tcx.sess.asm_arch.unwrap();
+        let reg_class = reg.reg_class();
+        let supported_tys = reg_class.supported_types(asm_arch);
+        let feature = match supported_tys.iter().find(|&&(t, _)| t == asm_ty) {
+            Some((_, feature)) => feature,
+            None => {
+                let msg = &format!("type `{}` cannot be used with this register class", ty);
+                let mut err = self.tcx.sess.struct_span_err(expr.span, msg);
+                let supported_tys: Vec<_> =
+                    supported_tys.iter().map(|(t, _)| t.to_string()).collect();
+                err.note(&format!(
+                    "register class `{}` supports these types: {}",
+                    reg_class.name(),
+                    supported_tys.join(", "),
+                ));
+                err.emit();
+                return Some(asm_ty);
+            }
+        };
+
+        // Check whether the selected type requires a target feature.
```

Comment:
Why is it necessary to check the output here, but the input during AST lowering? Cannot both be checked in the same location (either here or during lowering?)

Reply:
The set of types accepted by a register class can depend on what target features are available. See `kreg` for x86.

Reply:
The question is why we’re checking one set here and the other set there? Rather than everything in one location?

Reply:
I've added a comment explaining the difference between the two checks:

```rust
        // Check whether the selected type requires a target feature. Note that
        // this is different from the feature check we did earlier in AST
        // lowering. While AST lowering checked that this register class is
        // usable at all with the currently enabled features, some types may
        // only be usable with a register class when a certain feature is
        // enabled. We check this here since it depends on the results of typeck.
        //
        // Also note that this check isn't run when the operand type is never
        // (!). In that case we still need the earlier check in AST lowering to
        // verify that the register class is usable at all.
```